Basic barebones loop descriptor

davors72 opened the pull request.

Commit:
Local CI configuration update

 - Build master and testing-.* branches
 - Remove deploy/notification configurations

Commit:
Remove master from travis branches

s-perron committed:
Change IRContext::KillInst to delete instructions.

The current method of removing an instruction is to call ToNop.  The
problem with this is that it leaves around an instruction that later
passes will look at.  We should just delete the instruction.

In MemPass there is a utility routine called DCEInst.  It can delete
essentially any instruction, which can invalidate pointers now that they
are actually deleted.  The interface was changed to add a call back that
can be used to update any local data structures that contain
ir::Intruction*.

Commit:
Merge branch 'master' into testing_master

Commit:
Merge branch 'master' into testing_master

s-perron committed:
Compute value number when the value table is constructed.

Computing the value numbers on demand, as we do now, can lead to
different results depending on the order in which the users asks for
the value numbers.  To make things more stable, we compute them ahead
of time.

davors72 committed:
Dominator Tree Analysis (#3)

Support for dominator and post dominator analysis on ir::Functions. This patch contains a DominatorTree class for building the tree and DominatorAnalysis and DominatorAnalysisPass classes for interfacing and caching the built trees.

davors72 committed:
Merge branch 'SPV-512-Release' into testing-SPV-758

davors72 committed:
WIP Barebones commit

pierremoreau committed:
Opt: Remove unused lambda captures

Those are reported as errors by clang 5.0.0, due to the flags -Werror
and -Wunused-lambda-capture.

antiagainst committed:
Support outputting ANSI color escape sequences in library

Previously we required _PRINT to enable _COLOR, which forbids
outputting colored disassembly into a string in library.

This commit will allow library users to request enabling
ANSI color escape sequences.

antiagainst committed:
Fix Dref type check in validator

Dref should be of 32-bit scalar floating type.

Fixes https://github.com/KhronosGroup/SPIRV-Tools/issues/1012

antiagainst committed:
Appveyor: use ninja instead of MSBuild

antiagainst committed:
Travis CI: add clang-format check

Fixes https://github.com/KhronosGroup/SPIRV-Tools/issues/1029

antiagainst committed:
Travis: mark MacOS builds as optional

MacOS support in Travis is flakey right now. MacOS build jobs
take a long time to start.

We have extensive checks for Linux and Windows, which should give
us confidence.

dnovillo committed:
Re-format CONTRIBUTING.md - NFC.

I'm about to change this file.  I re-formatted it for easier editing.

dnovillo committed:
Add simplified instructions for pushing a reviewed PR.

davors72 committed:
Merge branch 'SPV-512-Release' into testing-SO-9-loop-descriptor

dwendelen committed:
Fixed typo that leaked to the binary

The typo was found by lintian when I was packaging glslang

davors72 committed:
Find loops in a given function.

davors72 committed:
Added the ability to track nested loops.

davors72 committed:
Removed appveyor and travis setup from branch.

Commit:
Add depth first iterator for trees

 - Add generic depth first iterator
 - Update the dominator tree to use this iterator instead of "randomly"
   iterate over the nodes

s-perron committed:
Add global redundancy elimination

Adds a pass that looks for redundant instruction in a function, and
removes them.  The algorithm is a hash table based value numbering
algorithm that traverses the dominator tree.

This pass removes completely redundant instructions, not partially
redundant ones.

s-perron committed:
DCEInst kill the same instruction twice.

In DCEInst, it is possible that the same instruction ends up in the
queue multiple times, if the same id is used multiple times in the
same instruction.

The solution is to keep the ids in a set, to ensure no duplication in
the list.

Commit:
Fix some of the known issues in image validation

Applied some of the spec clarifications made in conversation with
@johnkslang.

dnovillo committed:
Remove cfg_ field from SSAPropagator class - NFC.

When I moved the CFG into IRContext
(https://github.com/KhronosGroup/SPIRV-Tools/pull/1019), I forgot to
update SSAPropagator to stop requiring one.

Fixed with this patch.

dnovillo committed:
Update .appveyor.yml

dnovillo committed:
Change Release builds to Release with debug info on Linux

For Linux builds, we want to add debug information on the generated binary (for timing and profiling purposes).  This patch changes Release builds to RelWithDebInfo.

antiagainst committed:
Go back to root directory

antiagainst committed:
Do not zip and deploy for check formatting

Commit:
Add composite instruction validation pass

Validates instructions in the opcode range from OpVectorExtractDynamic
to OpTranspose.

antiagainst committed:
CMake: allow both SPIRV-Headers and spirv-headers

Fixes https://github.com/KhronosGroup/SPIRV-Tools/issues/1057

Commit:
Stop artifact stoarge with appveyor

dnovillo committed:
Sort flags in help message - NFC.

It's slightly easier to read flags from --help if they're sorted.

Commit:
Merge branch 'master' into testing_master

Naghasan commented on source/opt/loop_descriptor.cpp:
```diff
@@ -0,0 +1,123 @@
+namespace {
+
+// TODO: Move this into ir::Function
+static const spvtools::ir::BasicBlock* GetBasicBlock(
```
You should have this through the CFG no ?

Naghasan commented on source/opt/loop_descriptor.cpp:
```diff
@@ -0,0 +1,123 @@
+void LoopDescriptor::PopulateList(const ir::Function* f) {
+  ir::IRContext* context = f->GetParent()->context();
+
+  ir::CFG cfg{f->GetParent()};
```
Get it as a parameter

Naghasan commented on source/opt/loop_descriptor.cpp:
```diff
@@ -0,0 +1,123 @@
+  ir::CFG cfg{f->GetParent()};
+  opt::DominatorAnalysis* dom_analysis = context->GetDominatorAnalysis(f, cfg);
+
+  dom_analysis->DumpAsDot(std::cout);
```
debug code

Naghasan commented on source/opt/loop_descriptor.cpp:
```diff
@@ -0,0 +1,123 @@
+  // Function to find OpLoopMerge instructions inside the dominator tree.
+  auto find_merge_inst_in_dom_order =
+      [&loop_merge_inst](const DominatorTreeNode* node) {
+        ir::Instruction* merge_inst = node->bb_->GetLoopMergeInst();
```
not sure it is safe to assume the insn is present

davors72 replied:
It will return null if the tail isn't loop merge, it isn't safe if the BB doesn't have a tail but I don't think we could encounter that, right?

Naghasan replied:
The merge instruction might not be added (not sure we should rely on it).

In general, merge block may not be reachable (so won't be in the dom tree).

Naghasan commented on source/opt/loop_descriptor.cpp:
```diff
@@ -0,0 +1,123 @@
+  // However, this does not mean that dominance is implied by the order of
+  // loop_merge_inst you still need to check dominance between each block
+  // manually.
+  const DominatorTree& dom_tree = dom_analysis->GetDomTree();
```
analysis has a visit interface

Naghasan commented on source/opt/loop_descriptor.cpp:
```diff
@@ -0,0 +1,123 @@
+    // Otherwise, move through the loops in reverse order to check if this is a
+    // nested loop. If it isn't a nested loop this for will exit on the first
+    // iteration.
+    for (signed long i = loops_.size() - 1; i >= 0; --i) {
```
use rbegin()/rend()

Naghasan commented on source/opt/loop_descriptor.h:
```diff
@@ -0,0 +1,102 @@
+#define LIBSPIRV_OPT_LOOP_DESCRIPTORS_H_
+
+#include <cstdint>
+#include <map>
```
missing `#include <vector>`

make sure you run the cpplint.py

Naghasan commented on source/opt/loop_descriptor.h:
```diff
@@ -0,0 +1,102 @@
+      : loop_start_(begin),
+        loop_continue_(continue_target),
+        loop_merge_(merge_target),
+        is_nested_(is_nested){};
```
`const ir::BasicBlock` -> `ir::BasicBlock` (we want to operate on that loop).
This could be templated to create const/non-const Loop struct.

remove `;`

`is_nested` is unhelpful, Loop is a tree structure with a parent loop (if any) and children (a loop can have 1 to n loop)

Naghasan commented on source/opt/loop_descriptor.h:
```diff
@@ -0,0 +1,102 @@
+  std::vector<Loop*> nested_loops_;
+
+  // True if this loop is nested within another.
+  bool is_nested_;
```
Better to set a parent, IsNested becomes a null check

dnovillo committed:
Revert extraneous changes from commit 8ec62deb2.

Commit 8ec62deb2 merged the code from PR #810, but it also re-introduces
code that had been removed in #885.

This patch removes the (now superfluous code).

s-perron committed:
Kill names and decoration in inlining.

Currently when inlining a call, the name and decorations for the result of the
call is not deleted.  This should be changed.  Added a test for this as well.

This fixes issue #622.

Commit:
Merge branch 'master' into testing-SO-9-loop-descriptor

davors72 committed:
Updated in response to comments

Commit:
Merge branch 'master' into testing-SO-9-loop-descriptor

Commit:
Fix compilation issue after merge with master

davors72 committed:
Merge branch 'testing-SO-9-loop-descriptor' of https://github.com/codeplaysoftware/SPIRV-Tools into testing-SO-9-loop-descriptor

Commit:
Add operator[] to retrive dominator tree nodes from a basic block

s-perron committed:
Add option to spirv-opt to skip the validator.

The option --skip-validation is added.  When used it will skip the
validation step.

Also rearranged the help text to fix two options that are out of order.

Commit:
Store loop descriptors as unique pointers, change some variable names.

Loop descriptors were stored by value in a array,
which prevent the in place manipulation of the loop nests as
pointers might get invalidated.

greg-lunarg committed:
SingleStore: Support OpVariable Initialization

Treat an OpVariable with initialization as if it was an OpStore.
This fixes issue #1017.

dnovillo committed:
Add a new constant manager class.

This patch adds a new constant manager class to interface with
analysis::Constant.  The new constant manager lives in ir::IRContext
together with the type manager (analysis::TypeManager).

The new analysis::ConstantManager is used by the spec constant folder
and the constant propagator (in progress).

Another cleanup introduced by this patch removes the ID management from
the fold spec constant pass, and ir::IRContext and moves it to
ir::Module. SSA IDs were maintained by IRContext and Module.  That's
pointless and leads to mismatch IDs. Fixed by moving all the bookkeeping
to ir::Module.

pierremoreau committed:
Add external/SPIRV-Headers to .gitignore

b93c066b allowed for both SPIRV-Headers and spirv-headers as valid
paths, but only the latter is being ignored by Git.

Commit:
Merge branch 'master' into testing-SO-9-loop-descriptor

Commit:
Google C++ Style guide corrections

Commit:
Merge branch 'master' into testing_master

Commit:
Merge branch 'testing_master' into testing-SO-9-loop-descriptor
